subprojects/kotlin-dsl/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/BuildScriptCompileAvoidanceIntegrationTest.kt

```diff
@@ -0,0 +1,313 @@
+            """
+            buildscript {
+                dependencies { classpath(files("$jarPath")) }
+            }
+            $className().t4()
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("foo")
+
+        buildJarForBuildScriptClasspath(
+            """
+            public void t4() {
+                System.out.println("bar");
+            }
+            public void foo() {}
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("bar")
+    }
+
+    @ToBeFixedForConfigurationCache
+    @Test
+    fun `does not recompile buildscript on non ABI change in precompiled script plugin`() {
+        val pluginId = "my-plugin"
+        withPrecompiledScriptPluginInBuildSrc(
+            pluginId,
+            """
+                println("foo")
+            """
+        )
+        withBuildScript(
+            """
+                plugins {
+                    id("$pluginId")
+                }
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("foo")
+
+        withPrecompiledScriptPluginInBuildSrc(
+            pluginId,
+            """
+                println("bar")
+            """
+        )
+        configureProject().buildScriptNotCompiled().andOutputContains("bar")
+    }
+
+    @ToBeFixedForConfigurationCache
+    @Test
+    fun `recompiles buildscript when plugins applied from a precompiled plugin change`() {
+        val pluginId = "my-plugin"
+        withPrecompiledScriptPluginInBuildSrc(
+            pluginId,
+            """
+                plugins {
+                    id("java-library")
+                }
+                println("foo")
+            """
+        )
+        withBuildScript(
+            """
+                plugins {
+                    id("$pluginId")
+                }
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("foo")
+
+        withPrecompiledScriptPluginInBuildSrc(
+            pluginId,
+            """
+                plugins {
+                    id("java")
+                }
+                println("bar")
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("bar")
+    }
+
+    @ToBeFixedForConfigurationCache
+    @Test
+    fun `recompiles buildscript on inline function change in buildSrc`() {
+        val className = givenKotlinClassInBuildSrcContains(
+            """
+            inline fun t5() {
+                println("foo");
+            }
+            """
+        )
+        withBuildScript("$className().t5()")
+        configureProject().buildScriptCompiled().andOutputContains("foo")
+
+        givenKotlinClassInBuildSrcContains(
+            """
+            inline fun t5() {
+                println("bar");
+            }
+            """
+        )
+        configureProject().buildScriptCompiled().andOutputContains("bar")
+    }
```

Comment:
Looks like Kotlin's `inline` functions is where this approach falls short. This test fails - the build script does not get recompiled and it uses the inlined code from first compilation.

Reply:
Yeah, for that to work the classpath fingerprinter would need to be made aware of the Kotlin metadata for inline functions and include something like the bytecode of inline functions in the fingerprint.